src/Processors/QueryPlan/Optimizations/optimizeDirectReadFromTextIndex.cpp

```diff
@@ -37,6 +40,35 @@ String getNameWithoutAliases(const ActionsDAG::Node * node)
     return node->result_name;
 }
 
+const ActionsDAG::Node * replaceNodes(ActionsDAG & dag, const ActionsDAG::Node * node, const NodesReplacementMap & replacements)
+{
+    if (auto it = replacements.find(node); it != replacements.end())
+    {
+        return it->second;
+    }
+    else if (node->type == ActionsDAG::ActionType::ALIAS)
+    {
+        const auto * old_child = node->children[0];
+        const auto * new_child = replaceNodes(dag, old_child, replacements);
+
+        if (old_child != new_child)
+            return &dag.addAlias(*new_child, node->result_name);
+    }
+    else if (node->type == ActionsDAG::ActionType::FUNCTION)
+    {
+        auto old_children = node->children;
+        std::vector<const ActionsDAG::Node *> new_children;
+
+        for (const auto & child : old_children)
+            new_children.push_back(replaceNodes(dag, child, replacements));
+
+        if (new_children != old_children)
+            return &dag.addFunction(node->function_base, new_children, "");
+    }
+
+    return node;
```

Comment:
In the DAG there is only one list of node objects and the rest are pointers to the nodes in that list.

In that case if we iterate over the list and replace node info in place it is just simpler and faster. Don't need to store and search a `ReplacementMap`, don't need to iterate recursively over all the DAG and don't need to update any reference in a children because memory address won't change and the list entries are unique.

Did we have an issue with that approach?

Reply:
Technically, in-place modification was incorrect because we didn't update the `result_name` of the parent nodes; however, we should do this. It didn't cause bugs, but left a DAG in an inconsistent state. Also, I'd like to avoid modifying private fields of `ActionsDAG` inplace, and use public methods like `addInput` and `addFunction`. Iterating a DAG one extra time is not a big performance issue.